snapshot/storage/ql/ql.go

```diff
@@ -0,0 +1,362 @@
+
+	err := ms.doTransaction(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		for _, q := range schema {
+			_, err := tx.ExecContext(ctx, q)
+			if err != nil {
+				return errors.Wrap(err, "failed to exec create")
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return ms, nil
+}
+
+func (ms *qlMetastore) doTransaction(ctx context.Context, fn func(context.Context, *sql.Tx) error) error {
+	return ms.doDatabase(ctx, func(ctx context.Context, db *sql.DB) error {
+		tx, err := db.BeginTx(ctx, nil)
+		if err != nil {
+			return errors.Wrap(err, "failed to start transaction")
+		}
+
+		if err = fn(ctx, tx); err != nil {
+			if err1 := tx.Rollback(); err1 != nil {
+				err = errors.Wrapf(err, "rollback failure: %v", err1)
+			}
+			return err
+		}
+		if err := tx.Commit(); err != nil {
+			return errors.Wrap(err, "failed to commit")
+		}
+		return nil
+	})
+}
+
+func (ms *qlMetastore) doDatabase(ctx context.Context, fn func(context.Context, *sql.DB) error) error {
+	dbFile := filepath.Join(ms.dbfile)
+	// TODO: acquire lock on file
+
+	db, err := sql.Open("ql", dbFile)
+	if err != nil {
+		return errors.Wrap(err, "failed to open database file")
+	}
+	defer db.Close()
+
+	return fn(ctx, db)
+}
+
+func (ms *qlMetastore) Stat(ctx context.Context, key string) (snapshot.Info, error) {
+	var (
+		info   snapshot.Info
+		parent sql.NullString
+		active bool
+	)
+	info.Name = key
+
+	err := ms.doTransaction(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		err := tx.QueryRowContext(ctx, "SELECT b.Name, a.Active, a.Readonly FROM Snapshots as a LEFT JOIN (SELECT id() as ID, Name FROM Snapshots) as b ON a.Parent==b.ID WHERE a.Name=$1", key).Scan(&parent, &active, &info.Readonly)
+		if err != nil {
+			if err == sql.ErrNoRows {
+				// TODO: return snapshot package error
+				return errors.Errorf("snapshot %v not found", key)
+			}
+			return errors.Wrap(err, "select stat failed")
+		}
+		return nil
+	})
+	if err != nil {
+		return snapshot.Info{}, err
+	}
+	if parent.Valid {
+		info.Parent = parent.String
+	}
+	if active {
+		info.Kind = snapshot.KindActive
+	} else {
+		info.Kind = snapshot.KindCommitted
+	}
+
+	return info, nil
+}
+
+func (ms *qlMetastore) Walk(ctx context.Context, fn func(context.Context, snapshot.Info) error) error {
+	return ms.doDatabase(ctx, func(ctx context.Context, db *sql.DB) error {
+		rows, err := db.QueryContext(ctx, "SELECT a.Name, b.Name as Parent, a.Active, a.Readonly FROM Snapshots as a LEFT JOIN (SELECT id() as ID, Name FROM Snapshots) as b ON a.Parent==b.ID")
+		if err != nil {
+			return err
+		}
+
+		var (
+			info   snapshot.Info
+			active bool
+			parent sql.NullString
+		)
+		for rows.Next() {
+			if err := rows.Scan(&info.Name, &parent, &active, &info.Readonly); err != nil {
+				return err
+			}
+
+			info.Parent = parent.String
+			if active {
+				info.Kind = snapshot.KindActive
+			} else {
+				info.Kind = snapshot.KindCommitted
+
+			}
+
+			if err := fn(ctx, info); err != nil {
```

Comment:
Does `fn` inside `doDatabase` block anything?  It looks like this is executed while `ms.dbfile` is held open.  Should we read the results into a `[]snapshot.Info` and then call `fn` over each of them after `doDatabase` returns?